### PR TITLE
Basic function support

### PR DESCRIPTION
**Pull request by s0mark**
This is the result of my work in BSc project laboratory: basic function support implemented in Theta with locations stacks and instance mapping. It is enabled by default, function inlining can be turned on using the `--inline ON` flag. For details, see the [documentation](https://github.com/ftsrg/theta/files/9087273/function_support.pdf).

### CODE REVIEW

File: subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3SolverFactory.java [hunk 1]
```diff
@@ -45,10 +45,9 @@ private static void loadLibraries() {
 				System.loadLibrary("libz3java");
 				break;
 			case LINUX:
+			case MAC:
 				System.loadLibrary("z3java");
 				break;
-			default:
-				throw new UnsupportedOperationException("Operating system not supported.");
```
Comment (leventeBajczi) [inline subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3SolverFactory.java:51]: Nice work, but I think we should keep the default branch as a fallback

**Review by leventeBajczi — CHANGES_REQUESTED** [review]
Nicely done! 
I'll run a quick smoke test to see if the inlining/no-inlining functionality broke anything, but in the meantime, please fix the merge conflicts with master.
Are the new files under `lib` necessary for Theta to work on Mac? Is it enough (i.e., do all tests pass)?

**Comment by leventeBajczi** [comment]
I rebased it to the master branch so that I can run the smoke tests, see [function-merge](https://github.com/ftsrg/theta/tree/function-merge). The smoke tests ran fine against the ControlFlow section of the sv-comp benchmarks, but a test keeps failing in XcfaCliParseTest with the following log:

```
java.lang.NullPointerException: Cannot invoke "hu.bme.mit.theta.core.decl.VarDecl.getName()" because "var" is null
	at hu.bme.mit.theta.xcfa.model.XcfaProcedure.getAltVar(XcfaProcedure.java:245)
	at hu.bme.mit.theta.xcfa.model.XcfaProcedure.lambda$new$4(XcfaProcedure.java:88)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
	at hu.bme.mit.theta.xcfa.model.XcfaProcedure.<init>(XcfaProcedure.java:86)
	at hu.bme.mit.theta.xcfa.model.XcfaProcedure.duplicate(XcfaProcedure.java:228)
	at hu.bme.mit.theta.xcfa.model.XcfaProcess.lambda$new$3(XcfaProcess.java:82)
...
```  

Please take a look into its cause, and fix it if possible (it is more than probable that is is not your fault but the existing code I wrote).

Btw you can ignore the branch I created if you'd like to resolve the merge issues yourself, because I might have messed up there as well.

**Comment by s0mark** [comment]
> Are the new files under `lib` necessary for Theta to work on Mac? Is it enough (i.e., do all tests pass)?

Some symbolic links need to be created and an environmental variable has to be set to allow the libs to be found. With those done, all tests pass (by running the `test` task in the project's root).

> ... a test keeps failing in XcfaCliParseTest with the following log.

The issue was with (the previously disabled) `UnusedVarRemovalPass` once again, a method of it was called without running the pass. I [fixed](https://github.com/s0mark/theta/commit/41b132ec4e453c18da6bc1c747381825b627bd2e) the incompatibility of the pass with function support (and reenabled it), all tests in `XcfaCliParseTest` pass now.

> Btw you can ignore the branch I created if you'd like to resolve the merge issues yourself, because I might have messed up there as well.

It seemed fine, I [merged](https://github.com/s0mark/theta/commit/3f0e36f4bcb5cb76adba338d19b138ea4f8fe0b6) manually anyway just to make sure I didn't miss anything.

**Comment by leventeBajczi** [comment]
Thanks for your work!
Please bump the version number and then you can merge this PR. Don't forget to create a release as well.

**Pull request merged**